core/trino-main/src/main/java/io/trino/sql/planner/JsonPathEvaluator.java

```diff
@@ -0,0 +1,1187 @@
+            }
+            if (type.equals(REAL)) {
+                float value = intBitsToFloat((int) typedValue.getLongValue());
+                if (value > 0) {
+                    return typedValue;
+                }
+                return new TypedValue(type, floatToRawIntBits(Math.abs(value)));
+            }
+            if (type instanceof DecimalType) {
+                if (((DecimalType) type).isShort()) {
+                    long value = typedValue.getLongValue();
+                    if (value > 0) {
+                        return typedValue;
+                    }
+                    return new TypedValue(type, -value);
+                }
+                Int128 value = (Int128) typedValue.getObjectValue();
+                if (value.isNegative()) {
+                    Int128 result;
+                    try {
+                        result = DecimalOperators.Negation.negate((Int128) typedValue.getObjectValue());
+                    }
+                    catch (Exception e) {
+                        throw new PathEvaluationError(e);
+                    }
+                    return new TypedValue(type, result);
+                }
+                return typedValue;
+            }
+
+            throw itemTypeError("NUMBER", type.getDisplayName());
+        }
+
+        @Override
+        protected List<Object> visitIrArithmeticBinary(IrArithmeticBinary node, Context context)
+        {
+            List<Object> leftSequence = process(node.getLeft(), context);
+            List<Object> rightSequence = process(node.getRight(), context);
+
+            if (lax) {
+                leftSequence = unwrapArrays(leftSequence);
+                rightSequence = unwrapArrays(rightSequence);
+            }
+
+            if (leftSequence.size() != 1 || rightSequence.size() != 1) {
+                throw new PathEvaluationError("arithmetic binary expression requires singleton operands");
+            }
+
+            TypedValue left;
+            Object leftObject = getOnlyElement(leftSequence);
+            if (leftObject instanceof JsonNode) {
+                left = getNumericTypedValue((JsonNode) leftObject)
+                        .orElseThrow(() -> itemTypeError("NUMBER", ((JsonNode) leftObject).getNodeType().name()));
+            }
+            else {
+                left = (TypedValue) leftObject;
+            }
+
+            TypedValue right;
+            Object rightObject = getOnlyElement(rightSequence);
+            if (rightObject instanceof JsonNode) {
+                right = getNumericTypedValue((JsonNode) rightObject)
+                        .orElseThrow(() -> itemTypeError("NUMBER", ((JsonNode) rightObject).getNodeType().name()));
+            }
+            else {
+                right = (TypedValue) rightObject;
+            }
+
+            ResolvedOperatorAndCoercions operators = getOperators(OperatorType.valueOf(node.getOperator().name()), left.getType(), right.getType());
```

Comment:
Where is the operator being cached? Ideally, it should be attached to the `node` (via a map of node->operator) to avoid having to re-resolve it every time.

Reply:
The operators are cached in `CachingResolver`.